klippy/extras/load_cell_probe.py

```diff
@@ -0,0 +1,377 @@
+        # tare the sensor just before probing
+        # this uses pause(), requiring a print_time update
+        self._pause_and_tare()
+        # homing uses the toolhead last move time which gets special handling
+        # to significantly buffer print_time if the move queue has drained
+        toolhead = self.printer.lookup_object('toolhead')
+        print_time = toolhead.get_last_move_time()
+        # start collector, only used when probing
+        self._helper.notify_probe_start(print_time)
+        # start trsync
+        trigger_completion = self._dispatch.start(print_time)
+        clock = self._mcu.print_time_to_clock(print_time)
+        rest_ticks = self._mcu.seconds_to_clock(self._rest_time)
+        self._home_cmd.send([self._oid, self._dispatch.get_oid(),
+            mcu.MCU_trsync.REASON_ENDSTOP_HIT, self.REASON_SENSOR_ERROR, clock,
+            rest_ticks, WATCHDOG_MAX], reqclock=clock)
+        return trigger_completion
+
+    def home_wait(self, home_end_time):
+        self._dispatch.wait_end(home_end_time)
+        # trigger has happened, now to find out why...
+        res = self._dispatch.stop()
+        if res >= mcu.MCU_trsync.REASON_COMMS_TIMEOUT:
+            if res == mcu.MCU_trsync.REASON_COMMS_TIMEOUT:
+                raise self.printer.command_error(
+                    "Communication timeout during homing")
+            raise self.printer.command_error("Load cell sensor error")
+        if res != mcu.MCU_trsync.REASON_ENDSTOP_HIT:
+            return 0.
+        if self._mcu.is_fileoutput():
+            return home_end_time
+        self.last_trigger_time = self._clear_home()
+        return self.last_trigger_time
+
+    def query_endstop(self, print_time):
+        clock = self._mcu.print_time_to_clock(print_time)
+        if self._mcu.is_fileoutput():
+            return 0
+        params = self._query_cmd.send([self._oid], minclock=clock)
+        logging.info("Query Load Cell Endstop params returned: %s" % (params,))
+        if params['homing'] == 1:
+            return params['homing_triggered'] == 1
+        else:
+            return params['is_triggered'] == 1
+
+    ####
+    # Interface for ProbeEndstopWrapper
+    def _raise_probe(self):
+        toolhead = self.printer.lookup_object('toolhead')
+        start_pos = toolhead.get_position()
+        self.deactivate_gcode.run_gcode_from_command()
+        if toolhead.get_position()[:3] != start_pos[:3]:
+            raise self.printer.command_error(
+                "Toolhead moved during probe deactivate_gcode script")
+
+    def _lower_probe(self):
+        toolhead = self.printer.lookup_object('toolhead')
+        start_pos = toolhead.get_position()
+        self.activate_gcode.run_gcode_from_command()
+        if toolhead.get_position()[:3] != start_pos[:3]:
+            raise self.printer.command_error(
+                "Toolhead moved during probe activate_gcode script")
```

Comment:
It seems odd that the loadcell code is running arbitrary g-code here.  I'd like to understand the use case for that.  If there isn't a hard use-case I think we should not add it until we have a strong use-case for it.  In general, I'm concerned that running arbitrary g-code can make it difficult to predict behavior and difficult to make future changes.

Reply:
The use case is what Voron Tap is doing: stopping the probe from starting when the nozzle is too hot and will burn the PEI sheet. Reference: https://github.com/VoronDesign/Voron-Tap/blob/main/config/tap_klipper_instructions.md